ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/TransactionLogUpdatesConversions.scala

```diff
@@ -0,0 +1,423 @@
+                      workflowId = first.workflowId,
+                      effectiveAt = Some(instantToTimestamp(first.ledgerEffectiveTime)),
+                      events = flatEvents,
+                      offset = ApiOffset.toApiString(tx.offset),
+                      traceContext = None,
+                    )
+                  )
+              }
+        }
+        .getOrElse(Future.successful(None))
+    }
+
+    private val FlatTransactionPredicate =
+      (event: TransactionLogUpdate.Event, filter: FilterRelation) =>
+        if (filter.size == 1) {
+          val (party, templateIds) = filter.iterator.next()
+          if (templateIds.isEmpty)
+            event.flatEventWitnesses.contains(party)
+          else
+            // Single-party request, restricted to a set of template identifiers
+            event.flatEventWitnesses.contains(party) && templateIds.contains(event.templateId)
+        } else {
+          // Multi-party requests
+          // If no party requests specific template identifiers
+          val parties = filter.keySet
+          if (filter.forall(_._2.isEmpty))
+            event.flatEventWitnesses.intersect(parties.map(_.toString)).nonEmpty
+          else {
+            // If all parties request the same template identifier
+            val templateIds = filter.valuesIterator.flatten.toSet
+            if (filter.valuesIterator.forall(_ == templateIds)) {
+              event.flatEventWitnesses.intersect(parties.map(_.toString)).nonEmpty &&
+              templateIds.contains(event.templateId)
+            } else {
+              // If there are different template identifier but there are no wildcard parties
+              val partiesAndTemplateIds = Relation.flatten(filter).toSet
+              val wildcardParties = filter.filter(_._2.isEmpty).keySet
+              if (wildcardParties.isEmpty) {
+                partiesAndTemplateIds.exists { case (party, identifier) =>
+                  event.flatEventWitnesses.contains(party) && identifier == event.templateId
+                }
+              } else {
+                // If there are wildcard parties and different template identifiers
+                partiesAndTemplateIds.exists { case (party, identifier) =>
+                  event.flatEventWitnesses.contains(party) && identifier == event.templateId
+                } || event.flatEventWitnesses.intersect(wildcardParties.map(_.toString)).nonEmpty
+              }
+            }
+          }
+        }
+
+    private def permanent(events: Seq[TransactionLogUpdate.Event]): Set[ContractId] =
+      events.foldLeft(Set.empty[ContractId]) {
+        case (contractIds, event: TransactionLogUpdate.CreatedEvent) =>
+          contractIds + event.contractId
+        case (contractIds, event) if !contractIds.contains(event.contractId) =>
+          contractIds + event.contractId
+        case (contractIds, event) => contractIds - event.contractId
```

Comment:
I've explained above why the logic is correct. You are right that it might be very well to reverse engineer. I'll make the intent/logic more explanatory.

Reply:
Will adapt to the context for flat transactions only.